pydantic/dataclasses.py

```diff
@@ -321,11 +321,17 @@ def rebuild_dataclass(
                 types_namespace = {}
 
             types_namespace = _typing_extra.merge_cls_and_parent_ns(cls, types_namespace)
+
         return _pydantic_dataclasses.complete_dataclass(
             cls,
             _config.ConfigWrapper(cls.__pydantic_config__, check=False),
             raise_errors=raise_errors,
             types_namespace=types_namespace,
+            # We could provide a different config instead (with `'defer_build'` set to `True`)
+            # but this doesn't play well in some cases as unlike Pydantic models, Pydantic dataclasses
+            # can have the config coming from the `config` argument or from the `__pydantic_config__`
+            # property. So we use an explicit argument:
+            _force_build=True,
```

Comment:
For Pydantic models, we pass `{**config, "defer_build": False}` as the config, but for dataclasses this doesn't work well with this kind of scenario:

```python
@pydantic.dataclasses.dataclass(config=ConfigDict())
class Model:
    x: str
    __pydantic_config__ = ConfigDict(str_to_lower=True)
```

Hence the unfortunate new parameter.

Reply:
Hmm, I'd prefer we either disallow configs in both locations, or merge them with priority towards one.

That way, we could have one unified config for the model that we can patch with the `defer_build` when necessary.

Reply:
Should be able to move forward now that we've merged https://github.com/pydantic/pydantic/pull/10406

Reply:
Alright, I've updated this with the new changes!